File: oo-patterns/src/main/java/victor/training/oo/creational/singleton/AppConfiguration.java
package victor.training.oo.creational.singleton;

import static victor.training.oo.Helper.workSomeTime;

import java.io.IOException;
import java.util.Properties;

public class AppConfiguration {
	private static AppConfiguration INSTANCE; // SOLUTION

	private AppConfiguration() { // SOLUTION
//	public AppConfiguration() { // INITIAL	
		System.out.println("Creating singleton instance");
		properties = readConfiguration();
	}

	public static AppConfiguration getInstance() { // SOLUTION(
		if (INSTANCE == null) {
			INSTANCE = new AppConfiguration();
		}
		return INSTANCE;
	} // SOLUTION)

	private final Properties properties;

	private Properties readConfiguration() {
		System.out.println("Fetching encrypted configuration over HTTPS from Thailand");
		workSomeTime();
		System.out.println("Decrypting");
		workSomeTime();
		Properties props = new Properties();
		try {
			props.load(AppConfiguration.class.getResourceAsStream("dummy.properties"));
			System.out.println("Done");
			return props;
		} catch (IOException e) {
			throw new RuntimeException(e);
		}
	}
	
	public String getProperty(String propertyName) {
		return properties.getProperty(propertyName);
	}

}

File: oo-patterns-php/victor/training/oo/creational/singleton/AppConfiguration.php
<?php
/**
 * Created by IntelliJ IDEA.
 * User: VictorRentea
 * Date: 9/17/2017
 * Time: 7:42 PM
 */

namespace victor\training\oo\behavioral\singleton;


class AppConfiguration
{
	
    // SOLUTION (
    /* @var AppConfiguration */
    private static $INSTANCE;
	// SOLUTION )

    // public function __construct() // INITIAL 
    private function __construct() // SOLUTION 
	{
        printf("Creating singleton instance\n");
        $this->properties = $this->readConfiguration();
    }
	
	// SOLUTION (

    public static function getInstance(): AppConfiguration {
        if (static::$INSTANCE == null) {
            static::$INSTANCE = new AppConfiguration();
        }
        return static::$INSTANCE;
    }
    // SOLUTION )
    private $properties;

    private function readConfiguration(): array {
        printf("Fetching properties from Tahiti...\n");
        sleep(2);
        printf("Decrypting properties...\n");
        sleep(1);
        return parse_ini_file("props.ini");
    }

    public function getProperty(string $propertyName): string {
        return $this->properties[$propertyName];
    }
}